Defining built-ins in experimental analysis

cameel opened the pull request:
This PR removes hard-coded built-in type definitions and introduces the `__builtin()` construct for explicitly registering them.

It also renames the `integer` class to `Integer` to resolve the ambiguity with the `integer` type.

### To do:
- [x] Builtin-types
- [x] Decide how to parse `__builtin()`

cameel committed:
Add Builtin AST node

cameel committed:
Replace hard-coded types with __builtin()

cameel committed:
Reject duplicate built-in type definitions

cameel committed:
TypeSystemHelper: Add isPrimitiveType()

cameel committed:
Syntax test for built-in types in experimental Solidity.

cameel commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -1831,7 +1831,25 @@ ASTPointer<TypeDefinition> Parser::parseTypeDefinition()
+			expression = parseExpression();
+		else
+		{
+			expectToken(Token::Builtin);
+			expectToken(Token::LParen);
+
+			// TMP: Do I need to guard this?
+			//if (m_scanner->currentToken() == Token::StringLiteral)
+			expression = nodeFactory.createNode<BuiltinDefinition>(
+				std::make_shared<std::string>(m_scanner->currentLiteral()),
+				// TMP: Verify that it's the right location
+				m_scanner->currentLocation()
+			);
+
+			expectToken(Token::StringLiteral);
+			expectToken(Token::RParen);
```
Not sure if this is the best way to parse it:
- Maybe I should allow arbitrary arguments and only reject invalid ones in the analysis?
- Even if not, maybe the name of the builtin should be parsed into a `Literal` node? Currently I only expect a string and pass it as such to the `BuiltinDefinition` AST node.
    - If it's a literal, I can make AST visit it. Not sure if that would be useful though.

cameel commented on test/libsolidity/syntaxTests/experimental/builtin/builtin_type_definition.sol:
```diff
@@ -0,0 +1,88 @@
+type unit = __builtin("unit");
+
+type fun(T, U) = __builtin("fun");
+type pair(T, U) = __builtin("pair");
```
Since this is a built-in and the type variables will never be used for anything, maybe it would be better to omit them in the definition?

```solidity
type fun = __builtin("fun");
```

I could also do this, just to keep them connected to something:
```solidity
type fun(T, U) = __builtin("fun", T, U);
```
not sure if it makes sense though.

cameel commented on test/libsolidity/semanticTests/experimental/type_class.sol:
```diff
@@ -1,5 +1,8 @@
 pragma experimental solidity;
 
+type word = __builtin("word");
+type bool = __builtin("bool");
```
For type classes I'm going to introduce a very similar syntax:

```solidity
class integer = __builtin("integer");
```

Should not be unambiguous with full class definitions in the parser.

cameel replied:
Actually it might be ambiguous for operators in the future, at least if we ever plan to introduce `=` or `:`. For example for `=` we'd get this:

```solidity
class = = __builtin("=");
```

should add some extra tokens to avoid this?
```solidity
class { = } = __builtin("=");
```

```solidity
class operator = = __builtin("=");
```

I suspect the answer is "no" or at least "not yet" since we're going to adjust the syntax anyway.

ekpyron replied:
I don't think we'll ever (or at least in a very long time) *define* ``=`` and ``:`` - they're too basic and special. In principle, it was also merely a shortcut to reuse the operator tokens as names for the classes - usually they're given proper names anyways (like ``Equal`` or ```Plus``) and the special syntax is defined separately (although we won't be targetting definable syntax in a long time, so the syntax would be an effect of the builtin marker)

cameel replied:
Marking as resolved since we discussed it in depth on the channel and I'm going to deal with type classes in a separate PR after all.

cameel commented on libsolidity/experimental/analysis/TypeInference.cpp:
```diff
@@ -292,46 +276,10 @@ bool TypeInference::visit(InlineAssembly const& _inlineAssembly)
-				expressionAnnotation.type = m_typeSystem.type(*constructor, arguments);
-			else
-			{
-				TypeSystemHelpers helper{m_typeSystem};
-				expressionAnnotation.type =
-					helper.typeFunctionType(
-						helper.tupleType(arguments),
-						m_typeSystem.type(*constructor, arguments)
-					);
```
Do I still need this? I suspect I should now be doing this in `visit(Identifier)` when the identifier is a type name, but (at least in my simple tests) everything seems to be working just fine without it.

But maybe that's because I need more complex examples? Some expressions on types in variable declarations? Are those even implemented?

ekpyron replied:
All this becoming obsolete is kind of the point in making these types defined as usual - so it should work fine (and if it doesn't yet in all cases, we'll fix that elsewhere)

cameel replied:
But I meant this specific part of the function that I highlighted above. The assignment of a type annotation in term context is indeed handled when visiting the type definition now, but it does nothing in type context.

ekpyron replied:
I'm not sure I follow exactly :-). There is (may be) *some* difference between the old handling as ``ElementaryTypeNameExpression``s and the new one based on ``Identifier``s that refer to a type declaration (which are handled in ``handleIdentifierByReferencedDeclaration``).

But I think the new handling in ``Identifier`` should be the correct one. The old version may not even have worked for "function calls" to elementary type names in type expressions (not sure that even parsed), while the new one should.

In any case (without digging into it completely myself now), I think this is fine and if we notice that anything is missing later on, we can *then* fix it in the identifier visit.

cameel committed:
Distinct identifiers and tokens for integer type and integer class

- Using the same name causes syntax ambiguities. It's also not allowed for user-defined classes and types.

cameel commented:
Since we decided to do type classes in a separate PR, this is now ready.

ekpyron commented on libsolidity/ast/AST.h:
```diff
@@ -2596,6 +2596,35 @@ class TypeClassName: public ASTNode
 	std::variant<Token, ASTPointer<IdentifierPath>> m_name;
 };
 
+class BuiltinDefinition: public Expression
```
Actually, I'm not entirely sure about this name - is this really a "definition" :-)? We could just have plainly called it ``Builtin`` or ``BuiltinExpression`` - but let's not hold up the PR on this.

cameel replied:
Yeah, good point. `__builtin()` is not the whole type definition, just an expression. Your names are better. I'll rename it.

cameel replied:
Renamed to `Builtin`.

ekpyron commented on test/libsolidity/semanticTests/experimental/stub.sol:
```diff
@@ -1,5 +1,9 @@
 pragma experimental solidity;
 
+type word = __builtin("word");
+type bool = __builtin("bool");
```
For booleans, it's funny that we could actually define them completely as user-defined type, i.e.
```
type bool = word;
```
Or later, more precisely, as an ADT
```
data bool = true | false;
```

The interesting part about booleans is rather that we have "truthiness" and "falsiness" on them, allowing them to work in conditionals.
Which is a bit of a chicken and egg problem: how to define truthiness without a boolean to indicate whether something is truthy :-)? (For the ADT representation we could just mark the ``true`` constructor as truthy and the ``false`` constructor as falsy, but this would still take some doing) So let's keep it builtin for now, but down the road it may be interesting to think about how to improve upon that.

Funnily, we wouldn't even actually need "truthiness"/"falsiness", if we defined ``if``s and loops in-language, which would actually work with higher-order functions and the ability to call Solidity functions from Yul if they take only types that involve "Yul-types" - but for that we'd among others need lambdas and what-not, so let's don't go there for now :-).

We might keep them as builtin permanently anyways, especially if we, in the future, introduce a boolean type in Yul (which would be one option for addressing stuff like https://github.com/ethereum/solidity/issues/13750, which I moved to the optimizer ideas hackmd, but would also complicate things quite a bit, so also nothing we'd target soon)

ekpyron commented on libsolidity/experimental/analysis/TypeInference.cpp:
```diff
@@ -837,6 +779,20 @@ bool TypeInference::visit(TypeDefinition const& _typeDefinition)
+		ScopedSaveAndRestore expressionContext{m_expressionContext, ExpressionContext::Type};
+		_typeDefinition.typeExpression()->accept(*this);
+
+		if (!isBuiltIn)
```
If you move this ``if`` outside the other ``if`` you can only visit the ``typeExpression()`` if ``isBuiltIn`` is false, which would make the bogus ``visit(BuiltinDefinition const&)`` redundant and would give us the safety of rejecting builtins in unexpected context in ``visitNode``...

cameel replied:
Done.

ekpyron commented:
I'd say it'd be nice to address https://github.com/ethereum/solidity/pull/14566#discussion_r1331516030, but then feel free to just merge this!
